src/audio/pipeline/pipeline-stream.c

```diff
@@ -175,11 +175,28 @@ static int pipeline_trigger_list(struct pipeline *p, struct comp_dev *host, int
 	list_init(&walk_ctx.pipelines);
 
 	ret = walk_ctx.comp_func(host, NULL, &walk_ctx, host->direction);
-	if (ret < 0)
+	if (ret < 0) {
 		pipe_err(p, "pipeline_trigger_list(): ret = %d, host->comp.id = %u, cmd = %d",
 			 ret, dev_comp_id(host), cmd);
-	else
+	} else {
+		if (cmd == COMP_TRIGGER_PRE_START) {
+			struct list_item *list;
+			unsigned int n = 0, highest_prio;
+
+			/* Make sure the first pipeline has the highest priority */
+			list_for_item(list, &walk_ctx.pipelines) {
+				struct pipeline *p = list_item(list, struct pipeline,
+							       list);
+				if (n++) {
+					if (p->pipe_task->priority <= highest_prio)
+						p->pipe_task->priority = highest_prio + 1;
+				} else {
+					highest_prio = p->pipe_task->priority;
+				}
```

Comment:
we should be not altering priorities to start connected pipelines. We should really pre-walk the graph and detect all the connected pipelines for this trigger. i.e. we detec pipes 3, 7, 8 then we know we have to trigger those.

This graph pre walk could detect dependecy between connetced pipelines. i.e.
We could have a 
```c
struct pipeline {
...
/* if connected_source is not NULL it can be walked to trigger connected source pipelines */
struct comp_dev *connected_source; /* connected pipe component source */

/* if connected_sink is not NULL it can be walked to trigger connected sink pipelines */
struct comp_dev *connected_sink; /* connected pipe component sink */
};
```

Reply:
@lgirdwood this is evil I agree. But I don't think this can be solved outside of the scheduler. It isn't just about triggering. Consider this flow:
1. 2 connected pipelines are playing back audio:
1.1. the host pipeline is scheduled first, it copies data from the host down to the terminating buffer
1.2. the DAI pipeline is scheduled second and copies data from the buffer via mixer etc. down to the DAI
2. PAUSE is pressed:
2.1. the host pipeline returns COMPLETED and is removed from the scheduler list
2.2. the DAI pipeline with the mixer stays in the scheduler queue and continues playing back silence
3. RELEASE is pressed:
3.1. the host pipeline is re-scheduled.

Now, if that host pipeline has the same priority as the DAI pipeline, it will now be inserted into the queue after the DAI pipeline (first come first served). So their scheduling order will be reversed.

The only way to fix this *without* modifying priorities that I see is by keeping completed tasks on the queue until they're terminated by external code. But I don't like that very much either. Tasks should be able to terminate without external help. How about this: what if we add a new task return code, e.g. SUSPEND? If a task returns COMPLETED, it is removed from the queue fully and irreversibly, just like now. But if the task returns SUSPEND, it's not considered for scheduling any more but it's kept on the scheduler queue? Then pipeline tasks could always return SUSPEND - on both PAUSE and STOP and only be terminated and removed from the queue externally during pipeline tear-down?

Reply:
I think the simplest way to support connected pipelines is to 

1) copy 0s on underruns i.e source pipe is disabled, but sink pipe is running so we copy 0s (in frames) from source -> sink
2) overwrite older frames on overrun. i.e. sink pipe is disabled but source pipe is running so we copy new frames over oldest frames.

We already do something very similar for DSM with buffer flags where we allow underrun and overrun.

Reply:
@lgirdwood yes, that's also what is currently done with the mixer and what doesn't change with this PR. But what does that mean for this specific issue - how do you make sure, that the source pipe is rescheduled correctly upon RELEASE?

Reply:
with above change it does not matter if source or sink pipe is released first we should clear all buffers prior to restart. i.e. we will just play a few ms of 0s until the source pipe is ready.